src/CG/CG_elem.cpp

```diff
@@ -2113,6 +2113,829 @@ bool intersectSegmentPolygon( array3D const &P0, array3D const &P1, std::size_t
+        return normal;
+    }
+
+    if (nV == 3) {
+        //it's a regular triangle, can do it fast, without summation.
+        normal = crossProduct(V[2] - V[1], V[0] - V[1]);
+    } else {
+        //use vtkPolygon algorithm
+        int iV0(0), iV1(1), iV2(1);
+        for (std::size_t i = 0; i < nV; ++i) {
+            //update the vertices
+            iV0 = iV1;
+            iV1 = iV2;
+            iV2 = (i + 2) % nV;
+            normal += crossProduct(V[iV2] - V[iV1], V[iV0] - V[iV1]);
+        }
+    }
+
+    double normv = norm2(normal);
+    normv        = normv > std::numeric_limits<double>::min() ? normv : 1.0;
+
+    return normal / normv;
+}
+
+/*!
+ * Triangulation method: split the polygon in nV triangles adding the polygon centroid
+ * as auxiliary constitutive point.
+ * In case of non-convex polygon the method will create a non-compliant
+ * subdivision in subtriangles.
+ * When filling the sub-triangle connectivity, it will push the -1 label 
+ * as vertex index to recongnize that the auxiliary point Centroid is used.
+ * \param[in] nV Number  of Vertices of the polygon
+ * \param[in] V Pointer to Vertices list array. Must have size nV.
+ * \param[in] C Polygon centroid computed as the mean of the polygon vertices
+ * \param[in] subTrianglesCount number of generated triangles
+ * \param[in] subTrianglesConn connectivity of each triangle
+ * \return True in case of successfull triangulation, false otherwise
+ */
+bool triangulatePolygon(std::size_t nV, array3D const *V, array3D &C,
+                        int &subTrianglesCount, std::vector<int> &subTrianglesConn)
+{
+    //evaluate the centroid
+    C = {0.0, 0.0, 0.0};
+    for (std::size_t i = 0; i < nV; ++i) {
+        C += V[i];
+    }
+    C /= double(nV > 0 ? nV : 1);
+
+    //verify the set of vertices is defining a non-zero area polygon
+    array3D N = computePolygonNormal(nV, V);
+    if (norm2(N) <= std::numeric_limits<double>::min()) {
+        throw std::runtime_error ("List of vertices provided do not describe a polygon with area > 0.0. Cannot perform polygon triangulation.");
+        return false;
+    }
+
+    //run the triangulator core and update the sub-triangle structure.
+    subTrianglesCount = 0;
+    if (V) {
+        if (nV < 2 || V == nullptr) {
+            //line and vertex not allowed as polygon or if polygon is not provided properly
+            throw std::runtime_error ("Triangulation of polygon failed.");
+        } else if (nV == 3) {
+            //trivial trangulation, nothing to do.
+            subTrianglesCount = 1;
+            subTrianglesConn  = {{0, 1, 2}};
+        } else {
+            subTrianglesCount = nV;
+            subTrianglesConn.resize(3 * nV);
+
+            for (std::size_t i = 0; i < nV; ++i) {
+                subTrianglesConn[3 * i]     = i;
+                subTrianglesConn[3 * i + 1] = (i + 1) % nV;
+                subTrianglesConn[3 * i + 2] = -1;
```

Comment:
Are you triangulating the polygon? The module CG has functions to do that: subtriangleOfPolygon, polygonSubtriangleCount.

Reply:
This is a function written by Rocco. He didn't use the triangulation of bitpit on purpose, because it couldn't handle all the possible cases like concave polygons if I'm not mistaken.

Reply:
Both triangulations do the same thing: each sub-triangle is defined by two consecutive vertices and the centroid:
```
                subTrianglesConn[3 * i]     = i;
                subTrianglesConn[3 * i + 1] = (i + 1) % nV;
                subTrianglesConn[3 * i + 2] = -1;
```
with ```-1``` being replaced by the centroid of the polygon.

```
    V0 = {0., 0., 0.};
    for (std::size_t i = 0; i < nV; i++) {
        for (int d = 0; d < 3; d++) {
            V0[d] += V[i][d];
        }
    }
    for (int d = 0; d < 3; d++) {
        V0[d] /= nV;
    }
    V1 = V[triangle%nV];
    V2 = V[(triangle+1)%nV];
```
with ```V0``` being explicitly calculated as the centroid of the polygon.

This kind of sub-triangulation can fail for concave polygons.

Reply:
I simplified these methods. Now the triangulation is done under the hood and there's no need for calling polygonSubtriangleCount.